pydatastructs/linear_data_structures/algorithms.py

```diff
@@ -787,3 +789,168 @@ def longest_common_subsequence(seq1: OneDimensionalArray, seq2: OneDimensionalAr
                     check_mat[i][j] = check_mat[i][j-1]
 
     return OneDimensionalArray(seq1._dtype, check_mat[row][col][-1])
+
+'''Tim Sort Implementation'''
+
+def insertion_sort(array:Array, left,right=None)-> Array:
+    '''
+
+    This insertion sort function is required for implementation of timsort
+    In Insertion sort, The array is virtually split into a sorted and an
+    unsorted part. Values from the unsorted part are picked and placed at
+    the correct position in the sorted part.The first step in implementing
+    Timsort is modifying the implementation of insertion_sort() from before
+    : This modified implementation adds a couple of parameters, left and
+    right, that indicate which portion of the array should be sorted. This
+    allows the Timsort algorithm to sort a portion of the array in place.
+    Modifying the function instead of creating a new one means that it can
+    be reused for both insertion sort and Timsort.
+
+
+    '''
+    left=0
+    if right is None:
+        right = len(array) - 1
+
+    # Loop from the element indicated by
+    # `left` until the element indicated by `right`
+    for i in range(left + 1, right + 1):
+        # This is the element we want to position in its
+        # correct place
+        key_item = array[i]
+
+        # Initialize the variable that will be used to
+        # find the correct position of the element referenced
+        # by `key_item`
+        j = i - 1
+
+        # Run through the list of items (the left
+        # portion of the array) and find the correct position
+        # of the element referenced by `key_item`. Do this only
+        # if the `key_item` is smaller than its adjacent values.
+        while j >= left and array[j] > key_item:
+            # Shift the value one position to the left
+            # and reposition `j` to point to the next element
+            # (from right to left)
+            array[j + 1] = array[j]
+            j -= 1
+
+        # When you finish shifting the elements, position
+        # the `key_item` in its correct location
+        array[j + 1] = key_item
+
+    return array
+    """
+     Examples
+    ========
+
+    >>> from pydatastructs import OneDimensionalArray, insertion_sort
+    >>> arr = OneDimensionalArray(int,[3, 2, 1])
+    >>> insertion_sort(arr, 0)
+    >>> [arr[0], arr[1], arr[2]]
+    [1, 2, 3]
+
+    """
+
+def timsort(array : Array,**kwargs)-> Array:
+    """
+
+    The Timsort algorithm is considered a hybrid sorting algorithm because
+    it employs a best-of-both-worlds combination of insertion sort and
+    merge sort. The main characteristic of Timsort is that it takes
+    advantage of already-sorted elements that exist in most real-world
+    datasets.These are called natural runs. The algorithm then iterates
+    over the list, collecting the elements into runs and merging them into
+    a single sorted list.
+
+
+    """
+    start = kwargs.get('start', 0)
```

Comment:
start and end should be accepted from user. They shouldn't be kept optional

Reply:
In other functions, it is kept optional